### PR TITLE
Refactor: Using informal protocol instead of category override

### PR DESCRIPTION
**Pull request by cmds4410**
We can avoid overriding category methods by using an informal protocol with our `NSManagedObject` subclasses. 

#### Why do we care about overriding category methods? 
1. [According to Apple] (https://developer.apple.com/library/ios/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html#//apple_ref/doc/uid/TP40011210-CH6-SW4), `"If the name of a method declared in a category is the same as a method in the original class, or a method in another category on the same class (or even a superclass), the behavior is undefined as to which method implementation is used at runtime."`
While this seems to work for now, we **never** want the super class' implementation to be called instead of the implementation in the subclass. 
[In the past Apple has directly discouraged this.] (http://stackoverflow.com/questions/5272451/overriding-methods-using-categories-in-objective-c)
2. This is [explicitly] (https://developer.apple.com/library/ios/documentation/Swift/Conceptual/BuildingCocoaApps/InteractingWithObjective-CAPIs.html#//apple_ref/doc/uid/TP40014216-CH4-XID_32) not supported in Swift: `"You cannot use extensions to override existing methods or properties on Objective-C types."`

#### More importantly, how will this affect my existing projects?
Short answer: it won't. You do not need to change anything in your existing projects that make use of overriding methods like `rzv_primaryKey`. You **do not** need to `#import "RZVinylRecord.h"` or declare an interface like: 

```objective-c
#import "RZVinylRecord.h"
#import "RZVinylRip.h"

@interface RZPerson () <RZVinylRecord, RZVinylRip>

@end
```
But you can if you want.

#### What if a subclass does not implement one of the protocol methods?
All safety checks are encapsulated in the private methods like `+ (NSString *)rzv_safe_primaryKey;`
These safety methods check if self responds to selector, and if so calls it, otherwise returning a default value (just like before). **Every** protocol method has a corresponding safety method to be exclusively used **internally**. 
The mapping is as follows:

```objective-c
+ (RZCoreDataStack *)rzv_safe_coreDataStack; -> + (RZCoreDataStack *)rzv_coreDataStack;
+ (NSString *)rzv_safe_primaryKey; -> + (NSString *)rzv_primaryKey;
+ (NSPredicate *)rzv_safe_stalenessPredicate; ->  (NSPredicate *)rzv_stalenessPredicate;
```

**And for RZImport:**

```objective-c
+ (NSString *)rzv_safe_externalPrimaryKey; -> + (NSString *)rzv_externalPrimaryKey;
+ (BOOL)rzv_safe_shouldAlwaysCreateNewObjectOnImport; -> + (BOOL)rzv_shouldAlwaysCreateNewObjectOnImport;
```

#### Why Vinyl Rip?
[Vinyl ripping] (http://www.digitaltrends.com/home-theater/digitizing-your-record-collection/) is the process of digitizing your vinyl records into audio files. Feel free to rename it :)

*By the way, please let me know if having Swift support is not of interest in general.*

### CODE REVIEW

**Comment by ndonald2** [comment]
Thanks Connor, that's an excellent catch - I had never considered that this would be an issue. I guess I didn't realize that subclasses implementing overrides via categories were at risk for the method implementation race. I'll leave it to the RZ team to review, but this is a nice workaround.

**Comment by jkaufman** [comment]
I am strongly in favor of this change. Regardless of the risk and implications for Swift, adopting an informal protocol feels closer to canonical Cocoa. Thank you, @cmds4410.

**Comment by mgorbach** [comment]
Hi @cmds4410, thanks for this PR. Question for you: have you actually seen this happen in actual practice? It seems to me that we're not trying to override a category method in a category. We're not overriding a category method in the same class either. We're overriding a category method _in a subclass_. I'm not sure whether to read that as a problem based on the text below.
You solution actually seems quite nice and clean if we do have to do this. But is this absolutely necessary?

**Comment by cmds4410** [comment]
Hey thanks for reviewing @mgorbach (+ @ndonald2, and @jkaufman), and sorry for the delay in my response. What we are doing here, as the text above states, is declaring a method on a category of a subclass which is overrides a method declared in a category on the superclass. Unless I'm misreading the quote, this is explicitly undefined behavior. I have seen this in Swift, and it is definitely not guaranteed to work in the future. However I think it is equally important to consider that a protocol is more semantically accurate for this use case, and therefore makes it much easier for the consumer of this library to digest (IMO). Specifically, a protocol indicates *precisely* which methods should be implemented, whereas currently we rely on a pragma mark in the documentation to separate those which are intended to be overridden from those that are not. In this vein, it is perfectly fine to keep the more private implementations as categories, since they are not intended to be overridden by a subclass. As @jkaufman said, a protocol is closer to canonical cocoa.

**Comment by jkaufman** [comment]
@mgorbach Any objections to merging? While this may function in Objective-C, it is discouraged and even documented to _not_ work in Swift.

**Comment by mgorbach** [comment]
Yes, I think we should merge this. Unless anyone objects, I'll merge and release tomorrow.

**Comment by ndonald2** [comment]
:clap:

**Comment by ndonald2** [comment]
In case it wasn't obvious, I definitely approve. But to @jkaufman's point above - very little of the RZImport stuff will actually work in Swift since (last I checked) its runtime reflection was nowhere near as complete/documented as Objective-C's. So I guess it could work if you wrote your models in ObjC and bridged to swift... maybe.

**Comment by cmds4410** [comment]
Thanks @ndonald2! You're right about swift, and I can make another PR to the README that explains how to do it. Essentially you need to tag your NSManagedObject subclasses with @objc(ClassName) and use Objective-C types for all of your properties (NSString instead of String, etc). You don't need to actually write it in objc, just make it compile that way. Obviously that means you lose the value types advantages you get with Swift structs, but honestly they're not that useful in core data yet...hoping they make some Swift-targeted improvements this year.

**Comment by cmds4410** [comment]
And thanks again for reviewing @mgorbach

File: Classes/RZCoreDataStack.m [hunk 1]
```diff
@@ -29,9 +29,11 @@
 
 #import "RZCoreDataStack.h"
 #import "NSManagedObject+RZVinylRecord.h"
+#import "NSManagedobject+RZvinylRecord_private.h"
```
Comment (SpencerP) [inline Classes/RZCoreDataStack.m]: Very small note, but it looks like there's a typo here ("V" is not capitalized). #nth

Comment (cmds4410) [inline Classes/RZCoreDataStack.m]: Good catch

Comment (cmds4410) [inline Classes/RZCoreDataStack.m]: Hah it looks like that was always there!

Comment (SpencerP) [inline Classes/RZCoreDataStack.m]: Oops, well apparently it's not breaking anything!

**Comment by KingOfBrian** [comment]
Hey @cmds4410 -- any chance you can still aggravate this issue, and see if it's still an issue without this fix? The implementation is not very self evident, and having all these optional `rzv_` methods is a bit confusing.

I'm hoping that this was an issue with an early version of swift. This exact behavior is done in many places in UIKit, where over-riding UIViewController categories in subclasses (or in extensions in the swift world) is common place.

The first source you specify does seem a bit concerning, but is a bit hard to decipher given it's run-on nature. I could see the behavior of calling `super` being undefined with subclass categories, but we've been doing this exact thing all over in objective-c world and not seeing an issue. The second source `You cannot use extensions to override existing methods or properties on Objective-C types` just seems like a documentation bug at this point since it's considered best practice to override things in extensions.

Not a huge deal, just hoping to clarify what was going on.

**Comment by cmds4410** [comment]
Hey @KingOfBrian, I totally see what you're saying about the ambiguity added by introducing `rzv_` versions of all these methods. I can't say this with any authority, but you and @ndonald2 do seem to be correct that in Objective-C overriding methods in subclass categories doesn't run much or any risk in the method implementation race. That being said, the second source I cited was from the Swift documentation, and it remains there. I can do some more testing to see if this is still a problem in Swift 2, but even then the conclusion would be anecdotal and still in defiance of the docs. 

FWIW I have RZVinyl working in my 100% Swift project, and I'd be happy to share how that works or even make a PR to add an addendum to the README.

**Pull request merged**